// protractor-beautiful-reporter/examples/Gruntfile.js
module.exports = function(grunt) {

require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    shell: {
      "options": {
        stdout: true
      },
      "selenium": {
        command: './selenium/start',
        options: {
          stdout: false,
          async: true
        }
      },
      "protractor_install": {
        command: 'node ./node_modules/protractor/bin/webdriver-manager update'
      },
      "webdriver_start": {
          command: 'node ./node_modules/protractor/bin/webdriver-manager start'
      },
      "npm_install": {
        command: 'npm install'
      }
    },

    protractor: {
      options: {
        keepAlive: true,
        configFile: "protractor.jasmine2.conf.js"
      },
      singlerun: {},
      auto: {
        keepAlive: true,
        options: {
          args: {
            seleniumPort: 4444
          }
        }
      }
    }

  });


  grunt.registerTask('test:e2e', ['protractor:singlerun']);
  grunt.registerTask('install', ['update','shell:protractor_install']);
  grunt.registerTask('update', ['shell:npm_install']);

};
